src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Diagnostics;
+using Microsoft.VisualStudio.Notifications;
+using Microsoft.VisualStudio.Telemetry;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.ProjectSystem.PackageRestore;
+
+[Export(typeof(IPackageRestoreCycleDetector))]
+[method: ImportingConstructor]
+internal sealed class PackageRestoreCycleDetector(
+    UnconfiguredProject unconfiguredProject,
+    ITelemetryService telemetryService,
+    IProjectSystemOptions projectSystemOptions,
+    INonModalNotificationService userNotificationService)
+    : IPackageRestoreCycleDetector
+{
+    // This algorithm is to detect the pattern A -> B -> A -> B -> A in the most recent N values.
+    // To keep track of the most N recent values we are using a queue of fixed size, where:
+    //   The most recent value is inserted at the front in O(1) time, and the oldest value is removed at the 
+    // back in O(1) time.
+    //   The counter will keep track of how many times values have appeared in the queue consecutively.
+    // i.e.
+    // A -> B -> C -> A -> B -> D -> X -> Y -> X -> Y -> X -> Y -> X
```

Comment:
are we able to detect the cycle earlier? like this:
```
// A -> B -> C -> A -> B -> D -> X -> Y -> X -> Y -> X -> Y -> X
                                |--------------------|
```

Reply:
I can add a test for this. My sense is that this whole comment is factually incorrect.

Reply:
Added a test case and tweaked the comment a little.